Support only OS-agnostic file separator character

lread opened the pull request:
Explicitly support only forward slash as a file separator
character on Windows for input paths and patterns.

This allows for cross-OS linting.

Closes #6

lread committed:
Support only OS-agnostic file separator character

Explicitly support only forward slash as a file separator
character on Windows for input paths and patterns.

This allows for cross-OS linting.

Closes #6

lread commented:
@camsaul, took a minimal approach here.

This is technically a breaking change for Windows users.

Lemme know what you think.

camsaul commented on README.md:
```diff
@@ -72,6 +72,8 @@ Several options are currently supported:
 
 `:include-patterns` and `:exclude-patterns` accept either Strings or regex literals (regex literals cannot be embedded in EDN, so use string equivalents instead).
 
+Windows users: The backslash is not recognized as a file separator, use the forward slash `/` for the file separator character in paths and patterns.
```
Isn't the point of this PR that backslashes would automatically be replaced?

camsaul replied:
Nevermind. I don't think I understood the PR fully. So the patterns are specified with forward-slashes, and then when we get paths back programmatically we convert those to UNIX-style as well. Gotcha

camsaul commented on src/whitespace_linter.clj:
```diff
@@ -12,6 +12,8 @@
 
 (set! *warn-on-reflection* true)
 
+(def windows? (str/starts-with? (System/getProperty "os.name") "Windows"))
```
Maybe instead of checking `os.name` it makes more sense to just replace occurrences of `java.io.File/separator` with a backslash. Who knows, Microsoft might decide they want to change the name of the next version of Windows to winOS or something like that next year.

Or if somebody somewhere is using obscure OS that uses a different separator... better to not hardcode assumptions like these into the system when we can avoid them.

Or if WSL gets more deeply integrated into Windows in the future and WSL itself starts reporting `os.name` as Windows.

lread replied:
Thanks for the thought @camsaul!

Lemme rethink this generically...

We have asked the user to use forward slashes in paths and patterns.
When comparing paths we are scanning, we want to normalize to the forward slash separator.

I think I see whatcha mean.
We can instead do this: if `java.io.File/separator` is not forward slash: 
- then replace any `java.io.File/separator` in scanned paths with forward slash
- otherwise no need to change scanned paths

No longer any need to check `os.name`.

Cool, I'll give this change a go soon.

lread replied:
Ok @camsaul, have pushed the change.
Lemme know what you think.

lread replied:
Also manually tested with rewrite-clj on:
- macOS
- Windows
- Linux

camsaul reviewed (changes requested):
After thinking about this for a while I think this change makes the most sense. I was thinking maybe it would make sense to support automatic translation from Windows -> UNIX style separators in the regexes as well but that sounds error-prone and unnecessarily complicated.

The only change I would suggest is using `java.io.File/separator` directly rather than looking at `os.name` and making assumptions based on that

lread commented:
> After thinking about this for a while I think this change makes the most sense. I was thinking maybe it would make sense to support automatic translation from Windows -> UNIX style separators in the regexes as well but that sounds error-prone and unnecessarily complicated.

Yes, that was my thought too.

lread committed:
Review feedback: no longer relying on os.name

We can achieve the same goal, and more generically, without
checking "os.name" system property.
